overlord/snapstate/check_snap.go

```diff
@@ -613,10 +622,20 @@ var osutilEnsureUserGroup = osutil.EnsureUserGroup
 
 func validateSystemUsernames(si *snap.Info) error {
 	for _, user := range si.SystemUsernames {
-		if _, ok := supportedSystemUsernames[user.Name]; !ok {
+		systemUserName, ok := supportedSystemUsernames[user.Name]
+		if !ok {
 			return fmt.Errorf(`snap %q requires unsupported system username "%s"`, si.InstanceName(), user.Name)
 		}
 
+		if systemUserName.AllowedSnapIds != nil {
+			// Only certain snaps can use this user; let's check whether ours
+			// is one of these
+			if !strutil.ListContains(systemUserName.AllowedSnapIds, si.SnapName()) {
```

Comment:
this isn't the check we want, we want to check not the name of the snap, but the snap ID, i.e. 

```
$ snap info microk8s | grep snap-id
snap-id: EaXqgt1lyCaxKaQCU349mlodBkDCXRcg
```

so we need to be checking for `EaXqgt1lyCaxKaQCU349mlodBkDCXRcg` not `microk8s`.

In general though, I'm not really sure myself where we should:

1) store this information about what snap-ids are allowed to be used with what snap-ids
2) where to perform this check

I think that somewhere in check_snap.go is the right place to perform the check if we do end up storing the information here, since that's where we check other stuff about system-usernames and validateSystemUserNames, but if this information about which snap ids can use which system-usernames gets stored elsewhere I think this won't be applicable. 

I defer to @pedronis for an opinion on where to store this information

Reply:
> this isn't the check we want, we want to check not the name of the snap, but the snap ID, i.e.
> 
> ```
> $ snap info microk8s | grep snap-id
> snap-id: EaXqgt1lyCaxKaQCU349mlodBkDCXRcg
> ```

I don't get such a line when I run the same command, is it because the package I have was not installed from the store?

> so we need to be checking for `EaXqgt1lyCaxKaQCU349mlodBkDCXRcg` not `microk8s`.

But if this ID only comes from the store, maybe we could have a logic where the list of allowed snap is in the form `["name:<snap-name>", ...] ["id:<snap-id>", ...]` and
1. if the snap ID is found in the allowed list of ids (minus the "id:" prefix), then it's allowed, otherwise
2. if the snap has no ID but its name is found in the allowed list of names (minus the "name:" prefix), then it's allowed, or
3. in all other cases: denied.

Otherwise it would mean that sideloaded snaps would not be able to create their users, which would make debugging and testing harder.

> I think that somewhere in check_snap.go is the right place to perform the check if we do end up storing the information here, since that's where we check other stuff about system-usernames and validateSystemUserNames, but if this information about which snap ids can use which system-usernames gets stored elsewhere I think this won't be applicable.

The information will be coming from the snap store, this is just an *ad interim* solution while the store has not been updated to provide this information. Am I misunderstanding you, or are you suggesting for us to wait on a solution implemented in the store?

Reply:
Personally, I think that having review-tools prevent usage via blocking uploads to the store unless given specific permission is sufficient for this, but I think what @pedronis was going for was more that:

* if a snap is installed locally, it has no snap ID, and as such is allowed to use snap_microk8s by default (this matches how interface auto-connect assertions work, if a snap is installed locally, it is always allowed to install a plug, but no auto-connections happen)
* I suppose if we eventually know that this information will come from the store that is fine to keep it here, I just was more commenting that this file/package specifically doesn't seem like the best place to store policy information since it is kinda obscure and hard to find unless you know to go looking for it

Reply:
this would use si.SnapID then